### ISSUE TITLE
Navigation across folds

### ISSUE THREAD

**Issue by samson333** [XVimProject/XVim#827]
I found this issue #534 unsolved. Is it possible to map j and k to % when the cursor is at the fold?

**Comment by neybis** [comment]
I'm not sure if discussion should continue here or #534, but I am interested in investigating the implementation of this as well. My new job requires I use Visual Studio or Xcode. Being a linux-y vim guy, I went Xcode with XVim. That being said, I see some areas where this plug-in can be improved and I'm willing to invest some time here to make my workflow professionally more efficient.

In the past, I have made heavy use of folds and vim and this is pretty annoying for "folding" vim users.

Any ideas as to what this might take to implement?

**Comment by neybis** [comment]
So, I just noticed that if you use the arrow keys in normal mode XVim does **not** navigate across folds properly. However, if you go into insert mode and use the arrow keys there XVim **does** navigate across folds properly. Obviously, you can't test the home row key navigation in insert mode so that's kind of a moot point.

This is single handedly the most important item on my list of things I would like to see fixed. Especially now that I've issued the pull request for Vim folding commands (really brings the issue to the forefront).

**Comment by neybis** [comment]
I think I got it guys!

I found a snippet of code online that made a call down into Xcode to move the cursor rather than doing it itself. If I use this method, it skips over folds exactly as expected!!

I modified XVimNormalEvaluator.m to contain the following keybinding:
```
- (XVimEvaluator*)n{
	[[[[NSApplication sharedApplication] keyWindow] firstResponder] moveDown:self];
	return nil;
}
```

Now a `j` motion goes into a fold when it shouldn't and a `n` motion goes over a fold when it should. So how to we go about hooking this in to XVim?? Or is this approach just completely broken?

**Comment by JugglerShu** [comment]
@neybis 
Yes, you are right. In insert mode, XVim just passes the key handling to Xcode. That's why it works in insert mode.

To handle folding in normal mode properly you have to edit xvim_getMotionRange method is NSTextView+VimOperation.m and its related methods.
I think folding information is held by Xcode class that inherits from NSTextStorage. I once investigated this kind of thing but I forgot the detail.

About your XVimNormalEvaluator approach:
It is ok only for movements but Vim uses movements with operations like 'dj'.
As you can see another methods in XVimNormalEvaluator.m, they don't call Xcode methods directly.

**Comment by neybis** [comment]
Ahhh, yea... didn't even think about visual blocks and the many other commands you can do with vim motions. Hmmm... Well I think I'll start looking into where I can pull the folding information from Xcode. Getting that information will probably be half the battle (for me at least hahaha)!
